### PR TITLE
Use the DomElementSchemaRegistry for type checking in legacy mode

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@
  */
 
 import {BoundTarget, ConstantPool, CssSelector, DEFAULT_INTERPOLATION_CONFIG, DomElementSchemaRegistry, Expression, ExternalExpr, InterpolationConfig, LexerRange, ParseError, R3ComponentMetadata, R3TargetBinder, SelectorMatcher, Statement, TmplAstNode, WrappedNodeExpr, compileComponentFromMetadata, makeBindingParser, parseTemplate} from '@angular/compiler';
+import {SchemaMetadata} from '@angular/compiler/src/core';
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/annotations/src/component.ts]: This should not be a deep import

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts [hunk 2]
```diff
@@ -124,14 +124,51 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
...
+              ErrorCode.VALUE_HAS_WRONG_TYPE, rawExpr,
+              'NgModule.schemas must be an array of schemas');
+        }
+        const id = schemaRef.getIdentityIn(schemaRef.node.getSourceFile());
+        if (id === null || schemaRef.ownedByModuleGuess !== '@angular/core') {
+          throw new FatalDiagnosticError(
+              ErrorCode.VALUE_HAS_WRONG_TYPE, rawExpr,
+              'NgModule.schemas must be an array of schemas');
+        }
+        switch (id.text) {
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: Strictly speaking, the identifier could have been renamed. Wouldn't the `Reference.node` be a declaration from `@angular/core` typings and therefore offer a more robust way of identifying which schema to use?

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts [hunk 3]
```diff
@@ -124,14 +124,51 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
...
+          case 'CUSTOM_ELEMENTS_SCHEMA':
+            schemas.push(CUSTOM_ELEMENTS_SCHEMA);
+            break;
+          case 'NO_ERRORS_SCHEMA':
+            schemas.push(NO_ERRORS_SCHEMA);
+            break;
+          default:
+            throw new FatalDiagnosticError(
+                ErrorCode.VALUE_HAS_WRONG_TYPE, rawExpr,
+                `'${schemaRef.debugName}' is not a valid NgModule schema`);
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: Maybe list details on which schema's _are_ valid.

File: packages/compiler-cli/src/ngtsc/program.ts [hunk 4]
```diff
@@ -398,6 +399,7 @@ export class NgtscProgram implements api.Program {
         checkTypeOfBindings: true,
         checkTypeOfPipes: true,
         strictSafeNavigationTypes: true,
+        legacySchemaChecker: null,
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/program.ts]: Hmm, I wouldn't correlate `fullTemplateTypeCheck` with the legacy schema checking per se. If someone has `NgModule.schemas` configured, those will be completely ignored without any notice.

File: packages/compiler-cli/src/ngtsc/typecheck/src/api.ts [hunk 5]
```diff
@@ -102,4 +125,9 @@ export interface TypeCheckingConfig {
    * This is currently an unsupported feature.
    */
   checkQueries: false;
+
+  /**
+   *
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/typecheck/src/api.ts]: Oh noes, looks like we ran out of comments! 😱

File: packages/compiler-cli/src/ngtsc/typecheck/src/api.ts [hunk 6]
```diff
@@ -56,6 +66,19 @@ export interface TypeCtorMetadata {
...
+   * User-facing text of the error.
+   */
+  messageText: string;
+
+  code: ErrorCode;
+
+  span: ParseSourceSpan,
+}
+
+export type SchemaDiagnosticFn = (diagnostic: SchemaDiagnostic) => void;
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/typecheck/src/api.ts]: This doesn't look to be used anymore.

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts (hunk 1 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/component.ts]: Fixed!

File: packages/compiler-cli/src/ngtsc/typecheck/src/api.ts (hunk 5 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/api.ts]: 😆

File: packages/compiler-cli/src/ngtsc/typecheck/src/api.ts (hunk 6 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/api.ts]: Hahaha. Gone.

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: I added a comment explaining why it's safe :)

File: packages/compiler-cli/src/ngtsc/program.ts (hunk 4 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/program.ts]: Good point. Done.

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts (hunk 2 above)
Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: Ah, I see!

**Comment by alxhub** [comment]
Superseded by #32171.

**Pull request closed without merging**